consult-flymake.el

```diff
@@ -0,0 +1,104 @@
+;;; consult-flymake.el --- Provides the command `consult-flymake'  -*- lexical-binding: t; -*-
+
+;; Author: Daniel Mendler, Consult and Selectrum contributors
+;; Maintainer: Daniel Mendler
+;; Created: 2020
+;; License: GPL-3.0-or-later
+;; Version: 0.1
+;; Package-Requires: ((consult "0.1") (emacs "26.1"))
+;; Homepage: https://github.com/minad/consult
+
+;; This file is not part of GNU Emacs.
+
+;; This program is free software: you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Provides the command `consult-flymake'.  This is an extra package,
+;; to avoid loading Flymake.
+
+;;; Code:
+
+(require 'consult)
+(require 'flymake)
+
+(defcustom consult-preview-flymake t
+  "Enable Flymake preview during selection."
+  :type 'boolean
+  :group 'consult-preview)
+
+(defun consult-flymake--diag-type-name (diag)
+  "Return the type name of DIAG."
+  (let ((type (flymake-diagnostic-type diag)))
+    (format "%s"
+            (flymake--lookup-type-property
+             type 'flymake-type-name type))))
+
+(defun consult-flymake--candidates ()
+  "Return Flymake errors as alist."
+  (consult--forbid-minibuffer)
+  (unless (flymake-diagnostics)
+    (user-error "No Flymake diagnostics"))
+  (let* ((diagnostics (flymake-diagnostics))
+         (buffer-width (apply #'max (mapcar (lambda (x) (length (format "%s" (flymake-diagnostic-buffer x)))) diagnostics)))
+         (line-width (apply #'max (mapcar (lambda (x) (length (format "%s" (consult-flymake--diag-line x)))) diagnostics)))
+         (category-width (apply #'max (mapcar (lambda (x) (length (consult-flymake--diag-type-name x))) diagnostics)))
+         (fmt (format "%%-%ds %%-%ds %%-%ds %%s" buffer-width line-width category-width)))
```

Comment:
Iterate once over the diagnostics, remember category and line number?

Reply:
Feels like premature optimisation?

Reply:
There is a difference between optimization and trivially avoiding duplicate work. Why do we have to argue about everything here? This is not an efficient way of getting things done.

Reply:
See here on how this should have been written https://github.com/minad/consult/commit/4cd166da289f8bfe5e9bda6b6b7d047b6e16794b